README.md
# prankcall

Call a node.js generator with backoff and other customization

- Configure exponential backoff via [node-retry](https://github.com/tim-kos/node-retry#api) options
- Call the generator just once or repeatedly based on custom logic
- Optionally define another generator to receive results
- Sleep between successful calls
- Combine with modules like:
  - [co](https://github.com/visionmedia/co)
  - [co-request](https://github.com/leukhin/co-request)
- Observe retry events for logging

[![Build Status](https://travis-ci.org/codeactual/prankcall.png)](https://travis-ci.org/codeactual/prankcall)

## Example

> Use an async `producer` to generate input for an async `consumer`.
> `producer` might perform an HTTP request, while `consumer` stores the result in a database, job queue, etc.

```js
function *producer() {
  var data = yield producerWork();
  // ...
  return data;
}

function *consumer(dataFromProducer) {
  yield consumerWork(dataFromProducer);
  return shouldProducerRunAgain(); // {boolean}
}

co(function *() {
  var prankcall = require('prankcall').create();
  yield prankcall
    .sleep(1500) // Wait 1.5s after each successful `producer` call
    .retry({
      retries: 5,
      factor: 3,
      minTimeout: 1 * 1000,
      maxTimeout: 60 * 1000,
      randomize: true
    })
    .recv(consumer);
    .send(producer);
})();
```

See more [examples](test/lib/examples.js).

## Installation

### [NPM](https://npmjs.org/package/prankcall)

[![NPM](https://nodei.co/npm/prankcall.png?downloads=true)](https://nodei.co/npm/prankcall/)

    npm install prankcall

## API

- [Documentation](docs/Prankcall.md)
- [Events](docs/events.md)

## License

  MIT

## Tests

    npm test


CHANGELOG.md
# 0.2.1

- fix(es6): const violations

# 0.2.0

## breaking

- refactor(node): Migrate to ES6 features like `let` and `const`
  - Switch to `iojs` as only build target

## non-breaking

- refactor(extend): from migrate [assimlate](https://github.com/pluma0/assimilate) to [extend](https://github.com/justmoon/node-extend/)

# 0.1.3

- refactor(lib,test): remove isGeneratorFunction, works w/out those checks

# 0.1.2

- chore(npm): No changes, 0.1.1 publish appeared incomplete

# 0.1.1

- feat(recv): Support regular consumer functions
- docs(Prankcall): Mark #calcTimeouts as private for docs

# 0.1.0

- Initial API: `send`, `recv`, `sleep`, `retry`
- Initial events: `call`, `return`, `retry`


test/index.js
const T = module.exports = {};

const sinon = require('sinon');
const chai = require('chai');

T.should = chai.should();
chai.config.includeStack = true;
chai.use(require('sinon-chai'));

require('sinon-doublist')(sinon, 'mocha');

T.prankcall = require('../');


test/lib/examples.js
/*eslint func-names: 0, new-cap: 0, no-unused-expressions: 0, no-wrap-func: 0*/
'use strict';

const T = require('..');
const Prankcall = T.prankcall.Prankcall;
const sleep = Prankcall.sleep;
const request = require('co-request');

require('sinon-doublist-fs')('mocha');

describe('Prankcall - Examples', function() {
  beforeEach(function() {
    this.prankcall = T.prankcall.create();
  });

  function getWeatherFromYahoo(location) {
    return function *getWeatherFromYahooGen() {
      const response = yield request('http://query.yahooapis.com/v1/public/yql?q=select%20*%20from%20weather.forecast%20where%20location%20%3D%20' + location + '&format=json');
      return response.body;
    };
  }

  it('should demo single call', function *() {
    const fs = require('co-fs');
    let expectedJson;

    const filename = '/tmp/weather/latest.json';
    this.stubTree(filename);

    const producerGenFn = getWeatherFromYahoo(97204);

    const consumerGenFn = function *(dataFromProducer) {
      expectedJson = dataFromProducer;
      yield fs.writeFile(filename, dataFromProducer);
    };

    yield this.prankcall.recv(consumerGenFn).send(producerGenFn);

    (yield fs.readFile(filename)).toString().should.equal(expectedJson);
  });

  it('should demo producer dynamically stopped by consumer', function *() {
    const maxIter = 100;
    let curIter = 0;

    function *producer() {
      yield sleep(0);
      curIter++;
      return '';
    }

    function *consumer() {
      yield sleep(0);
      return curIter < maxIter;
    }

    this.prankcall.sleep(0);
    yield this.prankcall.recv(consumer).send(producer);

    curIter.should.equal(maxIter);
  });
});


test/lib/prankcall/index.js
/*eslint func-names: 0, new-cap: 0, no-unused-expressions: 0, no-wrap-func: 0*/
const T = require('../..');
const Prankcall = T.prankcall.Prankcall;
const sleep = Prankcall.sleep;
const retry = require('retry');

describe('Prankcall - Lib', function() {
  'use strict';

  beforeEach(function() {
    const test = this;

    // `this.send` will return one element at a time
    this.expectedCallReturn = ['one', 2, {three: true}];

    // To force retries
    this.sendErrMsg = 'test send error';
    this.sendErr = new Error(this.sendErrMsg);
    this.sendWithError = function *() {
      throw test.sendErr;
    };

    this.sleepTime = 1;
    this.fakeAsyncDelay = 3;

    this.defaultReceiveSpy = this.spy(Prankcall, 'defaultReceive');

    this.prankcall = T.prankcall.create();
    this.prankcall.set('sleep', this.sleepTime);

    this.sendSpy = this.spy();

    // Stand-in for something async like an HTTP request
    this.send = function *() {
      test.sendSpy();
      yield sleep(test.fakeAsyncDelay);
      return test.expectedCallReturn[test.recvCount];
    };

    // Store values passed to `this.recv`
    this.actualCallReturn = [];

    // Stand-in for something async, ex. write `this.send` results to DB
    this.recv = function *(callReturn) {
      yield sleep(test.fakeAsyncDelay);
      test.actualCallReturn.push(callReturn);
      return test.recvCount++ < test.expectedCallReturn.length - 1;
    };

    // To verify devs can omit the final `return` safely (no infinite loop)
    this.recvWithoutReturn = function *(callReturn) {
      test.actualCallReturn.push(callReturn);
      yield sleep(test.fakeAsyncDelay);
    };

    // Cursor used to read `this.expectedCallReturn` one element at a time
    this.recvCount = 0;

    this.useFakeBackoff = function() {
      test.stub(Prankcall, 'backoff', function() {
        return Prankcall.noOpCallback;
      });
    };

    this.nonGeneratorMultiIterRecv = function(callReturn) {
      test.actualCallReturn.push(callReturn);
      return test.recvCount++ < test.expectedCallReturn.length - 1;
    };

    this.nonGeneratorOneIterRecv = function(callReturn) {
      test.actualCallReturn.push(callReturn);
      return false;
    };

    this.nonGeneratorSend = function() {
      test.sendSpy();
      return test.expectedCallReturn[test.recvCount];
    };
  });

  it('should call #send until #recv returns false', function *() {
    this.prankcall.recv(this.recv);
    yield this.prankcall.send(this.send);
    this.actualCallReturn.should.deep.equal(this.expectedCallReturn);
  });

  it('should call #send only once if using default #recv', function *() {
    yield this.prankcall.send(this.send);
    this.sendSpy.should.have.been.calledOnce;
    this.defaultReceiveSpy.should.have.been.calledOnce;
  });

  it('should call #send only once if custom #recv returns undefined', function *() {
    this.prankcall.recv(this.recvWithoutReturn);
    yield this.prankcall.send(this.send);
    this.actualCallReturn.should.deep.equal([this.expectedCallReturn[0]]);
  });

  it('should propagate #send exception if retries exhausted', function *() {
    this.useFakeBackoff();

    let actual;
    try {
      yield this.prankcall.send(this.sendWithError);
    } catch (err) {
      actual = err;
    }
    actual.should.equal(this.sendErr);
  });

  it('should emit event: call', function *() {
    let actualStats;

    function onCall(stats) {
      actualStats = stats;
    }

    this.prankcall.on('call', onCall);

    yield this.prankcall.send(this.send);

    actualStats.should.deep.equal({calls: 0});
  });

  it('should emit event: return', function *() {
    let actualStats;
    let actualCallReturn;

    function onReturn(stats, callReturn) {
      actualStats = stats;
      actualCallReturn = callReturn;
    }

    this.prankcall.on('return', onReturn);

    yield this.prankcall.send(this.send);

    actualStats.should.deep.equal({calls: 1});
    actualCallReturn.should.equal(this.expectedCallReturn[0]);
  });

  it('should emit event: retry', function *() {
    this.useFakeBackoff();

    const actualDetails = [];

    function onRetry(details) {
      actualDetails.push(details);
    }

    this.prankcall.on('retry', onRetry);
    this.prankcall.retry({retries: 3});

    let caughtErr;
    try {
      yield this.prankcall.send(this.sendWithError);
    } catch (err) {
      caughtErr = err;
    }

    caughtErr.should.equal(this.sendErr);
    actualDetails.length.should.equal(3);
    actualDetails[0].should.deep.equal({
      backoffTime: 1000,
      currentRetries: 1,
      err: this.sendErr,
      remainRetries: 2,
      totalRetries: 1
    });
    actualDetails[1].should.deep.equal({
      backoffTime: 2000,
      currentRetries: 2,
      err: this.sendErr,
      remainRetries: 1,
      totalRetries: 2
    });
    actualDetails[2].should.deep.equal({
      backoffTime: 4000,
      currentRetries: 3,
      err: this.sendErr,
      remainRetries: 0,
      totalRetries: 3
    });
  });

  it('should recover from sparse #send exceptions', function *() {
    this.useFakeBackoff();

    const test = this;
    const successSequence = [true, false, false, true, false, false, true];
    const actualRetryDetails = [];

    function onRetry(details) {
      actualRetryDetails.push(details);
    }

    function *send() {
      if (!successSequence.shift()) {
        throw test.sendErr;
      }
      yield sleep(test.sleepTime);
      return test.expectedCallReturn[test.recvCount];
    }

    this.prankcall.on('retry', onRetry);
    this.prankcall.recv(this.recv);
    this.prankcall.retry({retries: 3});

    yield this.prankcall.send(send);

    actualRetryDetails.length.should.equal(4);
    this.actualCallReturn.should.deep.equal(this.expectedCallReturn);
  });

  it('should use default retry options', function *() {
    const spy = this.spy(retry, 'timeouts');

    yield this.prankcall.send(this.send);

    spy.should.have.been.calledWithExactly({});
  });

  it('should produce default timeout durations', function() {
    this.prankcall.calcTimeouts().should.deep.equal(
      [1000, 2000, 4000, 8000, 16000, 32000, 64000, 128000, 256000, 512000]
    );
  });

  it('should use custom retry options', function *() {
    const spy = this.spy(retry, 'timeouts');

    const custom = {
      timeout: 101,
      retries: 5,
      factor: 2.1,
      minTimeout: 2000,
      maxTimeout: 60000
    };
    this.prankcall.retry(custom);
    yield this.prankcall.send(this.send);

    spy.should.have.been.calledWithExactly(custom);
  });

  it('should use default sleep time', function *() {
    const spy = this.spy(GLOBAL, 'setTimeout');
    this.prankcall.recv(this.recv);
    yield this.prankcall.send(this.send);
    spy.args[0][1].should.equal(this.fakeAsyncDelay);
    spy.args[1][1].should.equal(this.fakeAsyncDelay);
    spy.args[2][1].should.equal(this.sleepTime);
    spy.args[3][1].should.equal(this.fakeAsyncDelay);
    spy.args[4][1].should.equal(this.fakeAsyncDelay);
    spy.args[5][1].should.equal(this.sleepTime);
    spy.args[6][1].should.equal(this.fakeAsyncDelay);
    spy.args[7][1].should.equal(this.fakeAsyncDelay);
  });

  it('should use custom sleep time', function *() {
    const spy = this.spy(GLOBAL, 'setTimeout');
    const custom = 33;
    this.prankcall.sleep(custom);
    this.prankcall.recv(this.recv);
    yield this.prankcall.send(this.send);
    spy.args[0][1].should.equal(this.fakeAsyncDelay); // send
    spy.args[1][1].should.equal(this.fakeAsyncDelay); // recv
    spy.args[2][1].should.equal(custom);              // sleep
    spy.args[3][1].should.equal(this.fakeAsyncDelay); // send
    spy.args[4][1].should.equal(this.fakeAsyncDelay); // recv
    spy.args[5][1].should.equal(custom);              // sleep
    spy.args[6][1].should.equal(this.fakeAsyncDelay); // send
    spy.args[7][1].should.equal(this.fakeAsyncDelay); // recv
  });

  it('should perform sleep', function() {
    const clock = this.sandbox.useFakeTimers();
    const ms = 86400;
    const thunk = sleep(ms);
    const spy = this.spy();
    thunk(spy);
    clock.tick(86399);
    spy.should.not.have.been.called;
    clock.tick(1);
    spy.should.have.been.called;
  });

  it('should pass call return values to non-generator #recv', function *() {
    const actualCallReturn = [];

    this.prankcall.recv(this.nonGeneratorMultiIterRecv.bind(this));
    yield this.prankcall.send(this.send);
    actualCallReturn.should.deep.equal(this.expectedCallReturn);
  });

  it('should call #send only once if custom non-generator #recv returns false', function *() {
    const actualCallReturn = [];

    this.prankcall.recv(this.recv);
    yield this.prankcall.send(this.send);
    actualCallReturn.should.deep.equal([this.expectedCallReturn[0]]);
  });

  it('should support non-generator used in #send', function *() {
    const actualCallReturn = [];

    this.prankcall.recv(this.recv);
    yield this.prankcall.send(this.nonGeneratorSend.bind(this));
    actualCallReturn.should.deep.equal(this.expectedCallReturn);
  });
});


lib/prankcall/index.js
/**
 * Call a node.js generator with backoff and other customization
 *
 * Licensed under MIT.
 * Copyright (c) 2014 David Smith <https://github.com/codeactual/>
 */

'use strict';

/**
 * Reference to Prankcall.
 */
exports.Prankcall = Prankcall;

/**
 * Create a new Prankcall.
 *
 * @return {object}
 */
exports.create = function create() { return new Prankcall(); };

/**
 * Extend Prankcall.prototype.
 *
 * @param {object} ext
 * @return {object} Merge result.
 */
exports.extend = function extendProto(ext) { return extend(Prankcall.prototype, ext); };

const events = require('events');
const retry = require('retry');
const util = require('util');

const extend = require('extend');
const configurable = require('configurable');

/**
 * Prankcall constructor.
 *
 * Usage:
 *
 *     const prankcall = require('prankcall').create();
 *     prankcall.retry({retries: 3}); // `retry` options
 *
 *     const send = function *() {
 *        // ... async read ...
 *        return data;
 *     };
 *     const recv = function *() {
 *        // ... async read ...
 *        return true; // keep reading
 *     };
 *
 *     yield prankcall.recv(recv).send(send);
 *
 * Configuration:
 *
 * - `{object} retry` Valid `retry` options
 *   - To disable: `prankcall.disable('retry')`
 *   - Default values below are from `retry`
 *   - `{number} [timeout=2000]`
 *   - `{number} [retries=10]`
 *   - `{number} [factor=2]`
 *   - `{number} [minTimeout=1000]`
 *   - `{number} [maxTimeout=Infinity]`
 *   - `{boolean} [randomize=false]`
 * - `{function *} [receive=none]` Optional consumer of each call's return value
 *   - `return true` to make another call
 *   - `return <falsey>` stop calling
 * - `{number} [sleep=1000]` Delay (in milliseconds) after successful call
 * @see tim-kos/node-retry https://github.com/tim-kos/node-retry/#api
 * @see Events https://github.com/codeactual/prankcall/blob/master/docs/events.md
 */
function Prankcall() {
  this.settings = {
    retry: {},
    recv: Prankcall.defaultReceive,
    sleep: 1000
  };
  events.EventEmitter.call(this);
}

util.inherits(Prankcall, events.EventEmitter);
configurable(Prankcall.prototype);

/**
 * Define a generator, or regular function, to receive the return values from
 * the generator passed to Prankcall.prototype.send.
 *
 * - Ex. to implement pagination by inspecting JSON responses from HTTP requests
 *   and then deciding whether to `send` should run again to fetch the next page.
 *   `recv` can signal `send` to continue/stop its loop by returning `true/false`.
 *
 * @param {function *} recv Generator or regular function.
 * - Receives the return value from `send` as the first argument.
 * @return {object} this
 */
Prankcall.prototype.recv = function recv(recv) {
  return this.set('recv', recv);
};

/**
 * Customize `retry` timeout options for backoff in `send`.
 *
 * @param {object} options Any valid `retry` timeout pairs
 * @return {object} this
 * @see tim-kos/node-retry https://github.com/tim-kos/node-retry/#api
 */
Prankcall.prototype.retry = function retry(options) {
  return this.set('retry', options);
};

/**
 * Sleep for a custom duration after each `send` iteration.
 *
 * @param {number} ms
 * @return {object} this
 */
Prankcall.prototype.sleep = function sleep(ms) {
  return this.set('sleep', ms);
};

/**
 * Yield to `sender` and retry after an exception.
 *
 * - By default, collect only one result from `sender`.
 *   - Return `true` from a Prankcall.prototype.receive generator to collect another.
 * - By default, backoff according to stock `retry` settings.
 *
 * @param {function *} sender
 */
Prankcall.prototype.send = function *send(sender) {
  let backoffTime;
  let callAgain = true;
  const callAgainSleep = this.get('sleep');
  let callReturn; // `return` value from `sender`
  const sleep = Prankcall.sleep;
  const stats = {
    calls: 0,
    currentRetries: 0,
    totalRetries: 0
  };
  let timeouts = this.calcTimeouts();
  const recv = this.get('recv');

  // Block executes on:
  // - 1st run
  // - Retry
  // - Receiver function returned `true`, see `this.get('recv')` below
  while (callAgain) {
    try {
      this.emit('call', {calls: stats.calls});

      callReturn = yield sender();

      stats.calls++;
      this.emit('return', {calls: stats.calls}, callReturn);
    } catch (err) {
      backoffTime = timeouts.shift();
      if (typeof backoffTime === 'undefined') { // Retries exhausted
        throw err;
      } else {
        stats.currentRetries++;
        stats.totalRetries++;
        this.emit(
          'retry',
          {
            backoffTime: backoffTime,
            remainRetries: timeouts.length,
            currentRetries: stats.currentRetries,
            err: err,
            totalRetries: stats.totalRetries
          }
        );
        yield Prankcall.backoff(backoffTime);
        continue;
      }
    }

    callAgain = !!(yield recv(callReturn));

    if (callAgain) {
      // Reset per-iteration states
      stats.currentRetries = 0;
      timeouts = this.calcTimeouts();

      this.emit('next', {sleepTime: callAgainSleep});

      if (callAgainSleep) {
        yield sleep(callAgainSleep);
      }
    }
  }
};

/**
 * Use `retry` to calculate timeout durations.
 *
 * @return {array} Ex. `[1000, 2000, 4000, ...]`
 * @api private
 */
Prankcall.prototype.calcTimeouts = function calcTimeouts() {
  const retryOpts = this.get('retry');
  return typeof retryOpts === 'object' ? retry.timeouts(retryOpts) : [];
};

/**
 * Err on the side of halting the generator after one call instead of
 * unexpectedly running it forever.
 */
Prankcall.defaultReceive = function defaultReceive() {
  return false;
};

Prankcall.noOpCallback = function noOpCallback(done) {
  done();
};

/**
 * Separate from Prankcall.sleep to allow separate test doubles.
 */
Prankcall.backoff = function backoff(ms) {
  return function prankCallBackoff(done) {
    setTimeout(done, ms);
  };
};

Prankcall.sleep = function sleep(ms) {
  return function prankCallSleep(done) {
    setTimeout(done, ms);
  };
};
